File: Combinations/Combinations.java
public class Solution {
    public List<List<Integer>> combine(int n, int k) {
        List<List<Integer>> ret = new ArrayList<List<Integer>>();
        if(n < k || n == 0 || k == 0)       return ret;
        
        
        if(k == 1){
            for(int i = 1; i <= n; ++i){
                List<Integer> element = new ArrayList<Integer>();
                element.add(i);
                ret.add(element);
            }
        }
        else{
            List<List<Integer>> prev = combine(n - 1, k - 1);
            for(List<Integer> l : prev){
                int last = l.get(k - 2);
                for(int i = last + 1; i <= n; ++i){
                    List<Integer> newElement = new ArrayList(l);
                    newElement.add(i);
                    ret.add(newElement);
                }
            }
        }
        
        return ret;
    }
}

File: Combinations/Combinations.cpp
class Solution {
public:

    vector<vector<int> > combine(int n, int k) {
        std::vector<std::vector<int> >  res;
        if(!n || !k)    return res;
        if(n < k)       return res;
        
        if(k == 1){
            for(unsigned int i = 0; i < n; ++i){
                std::vector<int> ret;
                ret.push_back(i + 1);
                res.push_back(ret);
            }
            return res;
        }
        else{
            for(unsigned int i = k; i <= n; ++i){
                std::vector<std::vector<int> > tmp = combine(i - 1, k - 1);
                //if(tmp.empty())     continue;
                for(unsigned int j = 0; j < tmp.size(); ++j){
                    tmp[j].push_back(i);
                    res.push_back(tmp[j]);
                }
            }
        }
        return res;
    }
};
